test/e2e/framework/service/jig.go

```diff
@@ -55,8 +57,20 @@ import (
 // NodePortRange should match whatever the default/configured range is
 var NodePortRange = utilnet.PortRange{Base: 30000, Size: 2768}
 
-// It is copied from "k8s.io/kubernetes/pkg/registry/core/service/portallocator"
-var errAllocated = errors.New("provided port is already allocated")
+// Initialize only once per test
+var (
+	staticPortAllocator = &staticPortRange{
+		nodeports: [87]bool{false},
+		baseport:  30000,
+		mu:        &sync.Mutex{},
+	}
+)
+
+type staticPortRange struct {
+	mu        *sync.Mutex
```

Comment:
Using a pointer to a mutex is usually wrong. Just use `sync.Mutex` here. (I'd probably use an embedded field here too. ie just `sync.Mutex` rather than `mu sync.Mutex`, and then use `s.Lock()` and `s.Unlock()` rather than `s.mu.Lock()` and `s.mu.Unlock()` later.)

Reply:
embedded fields for mutex are discouraged https://github.com/uber-go/guide/issues/127#issuecomment-876771151 , I could not find better references.

agree on the pointer thing, I missed that, it was not in the original suggestion https://github.com/kubernetes/kubernetes/pull/127153#discussion_r1751718361

Reply:
Embedding the mutex means that the `Lock` and `Unlock` methods become public methods of the type, which is definitely bad for exported types, but not always bad for simple private types.

But yeah, in this case the mutex is only locked/unlocked from within methods of the type, not by outside users of the type, so having it be private is better.